### PR TITLE
C#: Remove type args/params from generic type names in extractor

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR changes the extractor to remove `<,,,>` and `<T1,T2,T3>`-like suffixes from unbound and constructed generic types. With this change, the type names stored in the DB are always "undecorated" for generic types. Additionally, the QL library is adjusted to handle the name change (in unbound/constructed generic types, nullable types, and tuples).

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs [hunk 1]
```diff
@@ -485,18 +485,9 @@ private static void BuildNamedTypeDisplayName(this INamedTypeSymbol namedType, C
...
-                    {
-                        if (IsReallyBound(namedType))
-                        {
-                            p.BuildDisplayName(cx, trapFile);
-                        }
-                    });
-                trapFile.Write('>');
+                var args = string.Join(',', namedType.TypeArguments.Select(ta => ta.MetadataName));
+
+                cx.Extractor.Logger.Log(Util.Logging.Severity.Debug, $"Found generic type '{namedType.MetadataName}' with type arguments '{args}', skipping type arguments in type name.");
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs:490]: Is this debug statement really needed?

File: csharp/ql/src/semmle/code/csharp/Type.qll [hunk 2]
```diff
@@ -55,27 +55,8 @@ private predicate isObjectClass(Class c) { c instanceof ObjectType }
...
-  private string getQualifiedTypeArguments() {
-    result =
-      strictconcat(Type t, int i |
-        t = this.(ConstructedType).getTypeArgument(i)
-      |
-        t.getQualifiedName(), "," order by i
-      )
-  }
+  /** Gets the name of this type without `<...>` brackets, in case it is a generic type. */
+  string getNameWithoutBrackets() { types(this, _, result) }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Type.qll:59]: I suggest we deprecate `getNameWithoutBrackets`, and replace the body of `getUndecoratedName` with `types(this, _, result)`.

File: csharp/ql/src/semmle/code/csharp/Generics.qll [hunk 3]
```diff
@@ -360,21 +376,54 @@ class ConstructedType extends ValueOrRefType, ConstructedGeneric {
...
+        t = this.getTypeArgument(i)
       |
-        this.getTypeArgument(i).toString(), ", " order by i
+        t.getQualifiedName(), "," order by i
+      )
+  }
+
+  final override string toStringWithTypes() {
+    exists(string undecorated |
+      types(this, _, undecorated) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Generics.qll:404]: Use `GetUndecoratedName` instead.

File: csharp/ql/src/semmle/code/csharp/Generics.qll [hunk 4]
```diff
@@ -360,21 +376,54 @@ class ConstructedType extends ValueOrRefType, ConstructedGeneric {
...
+      result = undecorated + "<" + this.getTypeArgumentsString() + ">"
+    )
+  }
+
+  final override string getName() {
+    exists(string undecorated |
+      types(this, _, undecorated) and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/Generics.qll:411]: Use `GetUndecoratedName` instead.

**Review by hvitved — CHANGES_REQUESTED** [review]
We should also generate a DB upgrade script for adjusting the third column of the `types` relation.

**Comment by tamasvajk** [comment]
@hvitved I adjusted this PR based on your feedback. This is my first upgrade script, so please double check. I've ran the DB upgrade on a dummy DB, which was extracted with the previous version of the extractor, and I could query the generic types afterwards.

**Review by hvitved — DISMISSED** [review]
The upgrade script looks correct to me. I have approved the PR, but please do not merge before the `rc/3.2` release branch has been created, as there is no need for this change to go into the release.

**Comment by tamasvajk** [comment]
Diff job: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1235/

**Comment by tamasvajk** [comment]
Started a new diff job: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1242/

**Comment by tamasvajk** [comment]
new diff job to check useless upcast performance: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1245/

**Comment by tamasvajk** [comment]
diff job to check `getMember` inlining: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1246/

**Comment by tamasvajk** [comment]
@hvitved The diff jobs show better performance now. The first one shows some perf change in `DefaultToString` (+45s), and the second one with the `getMember` inlining improves it (-41s).

The failing test seems to be unrelated.

**Review by hvitved — APPROVED** [review]

**Pull request merged**